README.md
VisualStimulusToolbox 1.0.1
===========================
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.154061.svg)](https://doi.org/10.5281/zenodo.154061)

**VisualStimulusToolbox** is a lightweight MATLAB toolbox for generating, storing,
and plotting 2D visual stimuli commonly used in vision and neuroscience research,
such as sinusoidal gratings, plaids, random dot fields, and noise.

The toolbox allows for the easy creation, manipulation, plotting, and storing of visual stimuli such as drifting sinusoidal gratings, drifting plaids, drifting bars, random dot clouds, as well as their combinations. Every stimulus can be plotted, recorded to AVI, and stored to binary.

<div align="center">
  <img src="http://uci-carl.github.io/VisualStimulusToolbox/img/visualstimulus.jpg" style="width: 90%">
</div>

VisualStimulusToolbox was originally created to provide an easy way to important visual stimuli to the 
[CARLsim](http://www.socsci.uci.edu/~jkrichma/CARLsim) spiking network simulator. 
As of CARLsim 3.0, it is straightforward to convert VisualStimulus .dat files to spike trains that can serve
as input to CARLsim simulations. However, the toolbox can be used independently.

The toolbox is a lightweight alternative to the more comprehensive 
[Psychophysics](http://psychtoolbox.org) toolbox.

We use
[GitHub issues](https://github.com/UCI-CARL/VisualStimulusToolbox/issues)
for tracking requests and bugs.

If you use this code in a scientific contribution, please consider citing it as:
> Beyeler, M. "Visual Stimulus Toolbox: v1.0.0". Zenodo, June 22, 2016. doi:10.5281/zenodo.154061.

Or use the following bibtex:
```
@misc{visualstimulus,
  author       = {Michael Beyeler},
  title        = {Visual Stimulus Toolbox: v1.0.0},
  month        = {June},
  year         = {2016},
  doi          = {10.5281/zenodo.154061},
  url          = {http://dx.doi.org/10.5281/zenodo.154061},
  publisher    = {Zenodo}
}
```



## Installation

You can view and manage installed add-ons in MATLAB R2016a using the 
[Add-On Manager](http://www.mathworks.com/help/matlab/matlab_env/manage-your-add-ons.html).
To open the Add-On Manager, go to the **Home** tab, and select 
**Add-Ons** > **Manage Add-Ons**.

In older MATLAB versions, simply add the directory
**VisualStimulus/VisualStimulusToolbox** to your
[MATLAB path ](http://www.mathworks.com/help/matlab/ref/pathtool.html),
and you are good to go.

<div align="center">
<img src="http://www.mathworks.com/help/matlab/ref/set_path.png" alt="MATLAB pathtool" title="MATLAB pathool" width="60%"/>
</div>


## Getting Started

VisualStimulusToolbox provides a number of classes for creating, plotting,
and storing visual stimuli such as:
* `DotStim`: field of randomly drifting dots
* `GratingStim`: drifting sinusoidal grating
* `PlaidStim`: drifting plaid stimulus (composed of two sinusoidal gratings)
* `BarStim`: drifting bar stimulus
* `PictureStim`: stimulus made from one or several pictures (BMP, CUR, GIF, HDF, ICO, 
  JPEG, PBM, PCX, PGM, PNG, PPM, RAS, TIFF, XWD)
* `MovieStim`: stimulus made from one or several movies (AVI, MPG, MP4, M4V, MOV, WMV,
   MJ2, ASF, ASX)
* `CompoundStim`: stimulus made from a mixture of stimulus types listed
  above


#### Creating Your First Stimulus

A stimulus is intantiated by passing the desired stimulus height
and width (in pixels) to the constructor:

```Matlab
>> dot = DotStim([120 160])
  DotStim with properties:

                  width: 160
                 height: 120
               channels: 1
                 length: 0
                   stim: []
    supportedNoiseTypes: {'gaussian'  'localvar'  'poisson'  'salt & pepper'  'speckle'}
```

Frames can then be added using the method `add`, by specifying drift
direction (in degrees) and speed (in pixels/frame) as well as some other
stimulus-specific options (e.g., dot density, dot coherence, dot size,
etc.):

```Matlab
>> numFrames = 10;
>> dotSpeed = 1;
>> for dirDeg=(0:7)*45
	   dot.add(numFrames, dirDeg, dotSpeed);
   end
>> dot.plot;
```

This will create a stimulus made of a total of 80 frames, where dots drift
into one of eight directions (in 45 degree increments) for 10 frames each.

During plotting, key events can be used to pause, stop, and step through
the frames.
* Pressing `p` will pause plotting until another key is pressed.
* Pressing `s` will enter stepping mode, where the succeeding frame can be
  reached by pressing the right-arrow key, and the preceding frame can be
  reached by pressing the left-arrow key.
  Pressing `s` again will exit stepping mode.
* Pressing `q` will exit plotting.


Internally, the stimulus is stored as a 4D array
<height x width x channels x frames>.
For example, grayscale stimuli have one channel, and RGB stimuli have
three channels. The raw data array can also be accessed directly:
```Matlab
>> rawData4D = dot.stim;
```

Color stimuli can be created by passing a ColorSpec string to the
constructor:
```Matlab
>> dot = DotStim([120 160], 'r')
                  width: 160
                 height: 120
               channels: 3
                 length: 0
                   stim: []
    supportedNoiseTypes: {'gaussian'  'localvar'  'poisson'  'salt & pepper'  'speckle'}
```
Currently, the following color specs are supported: `'k'` (black), 
`'b'` (blue), `'g'` (green), `'c'` (cyan), `'r'` (red), `'m'` (magenta), `'y'` (yellow), 
and `'w'` (white).

A stimulus can also be converted to an AVI movie or stored as a binary
file (see below).


#### Manipulating an Existing Stimulus

Every stimulus type also comes with a number of handy helper methods:
* `clear`: Deletes all frames.
* `erase`: Deletes either a single frame or a list of frames.
* `popFront`: Deletes the first frame or number of frames.
* `popBack`: Deletes the last frame or number of frames.
* `rgb2gray`: Converts an RGB stimulus to a grayscale stimulus.
* `gray2rgb`: Converts a grayscale stimulus to RGB.
* `resize`: Resizes all frames by specifying either a scaling factor or
   a desired [height, width].
* `addNoise`: Adds noise to all existing frames. Supported noise types are
  given by variable `supportedNoiseTypes` and currently include Gaussian
  noise with constant mean, Gaussian white noise, Poisson noise, Salt &
  Pepper noise, and speckle (multiplicative) noise.


#### Combining Different Stimulus Types

Stimulus types can be combined to create compound stimuli.
Use optional input arguments to the constructor for fast stimulus
generation.
For example, to combine 10 frames of drifting sinusoidal grating
(120x180 pixels, red, drifting upwards and to the right at a 45 degree
angle)
with 20 frames of a drifting random dot cloud (240x360 pixels, blue,
drifting upwards at a 90 degree angle), use the following one-liner:
```
>> res = GratingStim([120 180],'r',10,45) + DotStim([240 360],'b',20,90)
```
If the two stimuli have distinct canvas dimensions, the second stimulus 
will be resized to match the first stimulus' [height width].
If any of the two stimuli have more than one color channel (e.g., RGB), 
the result will also have more than one color channel.


#### Recording AVI

Every stimulus type can be converted to an AVI movie using the `record`
method by specifying a desired file name and frame rate:
```Matlab
>> dot.record('myMovie.avi', 10); % 10 frames per second
```


#### Saving / Loading

Every stimulus type can be stored to a binary file, which can be loaded
at a later point:
```Matlab
>> dot.save('myBinaryStim.dat');
>> oldStim = dot.length;
>> newDot = DotStim;
>> newDot.load('myBinaryStim.dat');
>> assert(all(oldStim(:) == newDot.stim(:)))
```


## Acknowledgment
Some of this code is based on scripts initially authored by Timothy Saint
<saint@ncs.nyu.edu> and Eero P. Simoncelli <eero.simoncelli@nyu.edu>
at NYU for generating sinusoidal gratings, plaids, and dot clouds.

These scripts were released as part of the
[Motion Energy model](http://www.cns.nyu.edu/~lcv/MTmodel) in 2005,
which was released without stating any software license restrictions.
Their contributions are attributed at relevant places in the source code.


CHANGELOG
Legend:
 + added element
 - removed element
 * changed element

===========================================================================
Release 1.0
===========================================================================

---------------------
Build 1.0.1
---------------------

2017-11-28  Michael Beyeler <mbeyeler@uci.edu>
 - Removed outdated VisualStimulus.m

2016-06-22  Michael Beyeler <mbeyeler@uci.edu>
 * Improved CARLsim integration.
 - Removed outdated RELEASE_NOTES.

---------------------
Build 1.0.0
---------------------

2016-06-22  Michael Beyeler <mbeyeler@uci.edu>
 * Extend PictureStim and MovieStim to being able to combine multiple
   elements.

2016-06-21  Michael Beyeler <mbeyeler@uci.edu>
 + Add color option to all stimulus types.
 * Extend source code documentation.

2016-06-16  Michael Beyeler <mbeyeler@uci.edu>
 + Add recording function for storing stimuli as AVI files.
 + Add save-to-file function for storing stimuli as raw binary data.
 + Add PictureStim for creating a stimulus from a picture.
 + Add MovieStim for creating a stimulus from a movie.
 + Add noise option for adding noise to specified frames.
 + Add resize function for resizing stimuli dimensions.

2016-06-08  Michael Beyeler <mbeyeler@uci.edu>
 + Add interactive plot mode for stepping forward/backward through frames.

2016-06-07  Michael Beyeler <mbeyeler@uci.edu>
 + Add DotStim for creating random dot clouds.

2016-06-04  Michael Beyeler <mbeyeler@uci.edu>
 + Add GratingStim for creating sinusoidal gratings.

2016-05-20  Michael Beyeler <mbeyeler@uci.edu>
 + Add barebone structure for base/derived classes
 + Add BarStim for creating a drifting bar stimulus.
 + Add Toolbox project


===========================================================================
Release 0.1
===========================================================================

2014-10-14	Michael Beyeler <mbeyeler@uci.edu>
 + Add code base from CARLsim repo.


VisualStimulus/DotStim.m
classdef DotStim < BaseStim
    methods (Access = public)
        %% Public Methods
        function obj = DotStim(dimHW, colorChar, length, dir, speed, ....
                density, coherence, dotRadius)
			% dot = DotStim(dimHW, colorChar), for 1-by-2 vector DIMHW and
			% single character COLOR, creates a drifting random dot cloud
			% with dimensions [height width] and color 'k' (black), 'b' 
            % (blue), 'g' (green) 'c' (cyan), 'r' (red), 'm' (magenta), 
            % 'y' (yellow), or 'w' (white).
			%
			% dot = DotStim(dimHW), for vector DIMHW, creates a grayscale
            % drifting random dot cloud with dimensions [height width].
			%
			% dot = DotStim, creates an empty random dot cloud container.
			% Call dot.load(fileName) to load a previously saved dot cloud.
			%
			% dot = DotStim(dimHW, colorChar, length, dir, speed, density,
			% coherence, dotRadius), creates a drifting random dot cloud of 
            % LENGTH frames, where a fraction COHERENCE drifts coherently 
            % into direction DIR (between 0 and 360 degrees) with speed
            % SPEED (in pixels per frame). The remaining dots all drift
            % into random directions. DENSITY (between 0 and 1) will create
            % roughly DENSITY*height*width dots. DOTRADIUS indicates the
            % size of the dots, with DOTRADIUS<0 indicating single-pixel
            % size.
			%
			% DIMHW      - 1-by-2 stimulus dimensions: [height width].
			%              Default: [0 0].
			% COLORCHAR  - ColorSpec: 'k' (black), 'b' (blue), 'g' (green),
			%              'c' (cyan), 'r' (red), 'm' (magenta), 'y' 
			%              (yellow), and 'w' (white).
			%              Default: 'w'.
			% LENGTH     - Number of frames to create.
			%              Default: 0.
			% DIR        - Drifting directions in degrees (where
			%              0=rightward, 90=upward, 180=leftward,
			%              270=downward).
			%              Default: 0.
			% SPEED       - 1-by-2 stimulus frequency: [spatFreq tempFreq].
			%              Spatial frequency in cycles/pixels, temporal
			%              frequency in cycles/frame.
			%              Default: [0.1 0.1]
			% DENSITY    - Dot density (between 0 and 1). This will create
			%              roughly DENSITY*height*width dots.
			%              Default: 0.1
			% COHERENCE  - Fraction of dots drifting coherently in a given
			%              direction (between 0 and 1). The remaining dots
			%              drift in random directions.
			%              Default: 1.
			% DOTRADIUS  - Dot radius (in pixels). If DOTRADIUS<0, the dots
			%              will have single-pixel size. If DOTRADIUS>0, the
			%              dots will be Gaussian blobs with Gaussian width
			%              sigma=0.5*DOTRADIUS.
			%              Default: -1.
			if nargin == 0
				obj.height = 0;
				obj.width = 0;
			else
				obj.height = dimHW(1);
				obj.width = dimHW(2);
			end
			if nargin<2,colorChar='w';end
			
			% initialize default parameter values
			obj.initDefaultParams();

			% determine whether stimulus should be grayscale or color
			if strcmpi(colorChar, 'w') || strcmpi(colorChar, 'k')
				obj.channels = 1;
			else
				obj.channels = 3;
			end
			obj.colorChar = colorChar;
			obj.colorVec = obj.char2rgb(colorChar);

			% optionally, create stimulus in the same line of code
            if nargin >= 3
                if nargin<4,dir=obj.dotDirection;end
                if nargin<5,speed=obj.dotSpeed;end
                if nargin<6,density=obj.dotDensity;end
                if nargin<7,coherence=obj.dotCoherence;end
                if nargin<8,dotRadius=obj.dotRadius;end
                
				validateattributes(length, {'numeric'}, ...
					{'integer','nonnegative'}, 'DotStim', 'length')
				validateattributes(dir, {'numeric'}, ...
					{'real','>=',0,'<=',360}, 'DotStim', 'dotDirection')
				validateattributes(speed, {'numeric'}, ...
					{'vector', 'real', 'nonnegative'}, 'DotStim', ...
					'dotSpeed')
				validateattributes(density, {'numeric'}, ...
					{'real','>',0,'<=',1}, 'DotStim', 'contr')
                validateattributes(coherence, {'numeric'}, ...
                    {'real','>=',0,'<=',1}, 'DotStim', 'dotCoherence')
                validateattributes(dotRadius, {'numeric'}, ...
                    {'real'}, 'DotStim', 'dotRadius')
				
                % shortcut to create and add
                obj.add(length, dir, speed, density,...
                    coherence, dotRadius);
            end
        end
        
        function add(obj, length, dir, speed, density, coherence, ...
                dotRadius)
            % dot.add(length, type, dir, speed, density, coherence, radius)
            % appends LENGTH frames of a random dot cloud to the existing
            % stimulus, where a fraction COHERENCE of all dots drift into
            % direction DIR (between 0 and 360 degrees) with speed SPEED
            % (in pixels per frame). Dot density (between 0 and 1)
            % determines how many dots are created (# dots is roughly
            % DENSITY*height*width). The size of the dots is given by
            % DOTRADIUS.
            %
            % dot.add(length) appends LENGTH frames of a random dot cloud
            % to the existing stimulus, using the dot cloud specifications
            % from the constructor call.
            %
			% Warning: You cannot add frames to 0-by-0 sized stimulus. Use
			% the empty constructor only to load stimuli.
			%
            % This method uses a script initially authored by Timothy Saint 
            % and Eero P. Simoncelli at NYU.
            %
			% DIMHW      - 1-by-2 stimulus dimensions: [height width].
			%              Default: [0 0].
			% COLORCHAR  - ColorSpec: 'k' (black), 'b' (blue), 'g' (green),
			%              'c' (cyan), 'r' (red), 'm' (magenta), 'y' 
			%              (yellow), and 'w' (white).
			%              Default: 'w'.
			% LENGTH     - Number of frames to create.
			%              Default: 0.
			% DIR        - Drifting directions in degrees (where
			%              0=rightward, 90=upward, 180=leftward,
			%              270=downward).
			%              Default: 0.
			% SPEED       - 1-by-2 stimulus frequency: [spatFreq tempFreq].
			%              Spatial frequency in cycles/pixels, temporal
			%              frequency in cycles/frame.
			%              Default: [0.1 0.1]
			% DENSITY    - Dot density (between 0 and 1). This will create
			%              roughly DENSITY*height*width dots.
			%              Default: 0.1
			% COHERENCE  - Fraction of dots drifting coherently in a given
			%              direction (between 0 and 1). The remaining dots
			%              drift in random directions.
			%              Default: 1.
			% DOTRADIUS  - Dot radius (in pixels). If DOTRADIUS<0, the dots
			%              will have single-pixel size. If DOTRADIUS>0, the
			%              dots will be Gaussian blobs with Gaussian width
			%              sigma=0.5*DOTRADIUS.
			%              Default: -1.
            if nargin<3,dir=obj.dotDirection;end
            if nargin<4,speed=obj.dotSpeed;end
            if nargin<5,density=obj.dotDensity;end
            if nargin<6,coherence=obj.dotCoherence;end
            if nargin<7,dotRadius=obj.dotRadius;end
            validateattributes(length, {'numeric'}, ...
                {'integer','positive'}, 'add', 'length')
            validateattributes(dir, {'numeric'}, {'real'}, 'add', ...
                'direction')
            validateattributes(speed, {'numeric'}, ...
                {'integer','nonnegative'}, 'add', 'speed')
            validateattributes(density, {'numeric'}, ...
                {'real','>=',0,'<=',1}, 'add', 'density')
            validateattributes(coherence, {'numeric'}, ...
                {'real','>=',0,'<=',1}, 'add', 'coherence')
            validateattributes(dotRadius,{'numeric'}, {'real'}, 'add', ...
                'radius')
            
			if obj.height == 0 || obj.width == 0
				msgId = [obj.baseMsgId ':InvalidStimulusDimensions'];
				msg = 'Cannot add frames to 0-by-0 stimulus.';
				error(msgId, msg)
			end
			
			if density<0
                if dotRadius<0
                    density = 0.1;
                else
                    density = 0.3/(pi*dotRadius^2);
                end
            end
            
            % make the large dot for sampling
            rLargeDot = linspace(-1.5*dotRadius, 1.5*dotRadius, ...
                ceil(3*dotRadius*obj.sampleFactor));
            largeDot = exp(-rLargeDot.^2 ./ (2.*(dotRadius/2)^2));
            largeDot = largeDot'*largeDot;
            [rLargeDot, cLargeDot] = meshgrid(rLargeDot, rLargeDot);
            
            % There is a buffer area around the image so that we don't 
            % have to worry about getting wraparounds exactly right.
            % This buffer is twice the size of a dot diameter.
            if dotRadius > 0
                bufferSize = 2*ceil(dotRadius*3);
            else
                bufferSize = 5;
            end
            
            % the 'frame' is the field across which the dots drift. The
            % final output of this this function will consist of
            % 'snapshots' of this frame without buffer.
            rowFrame = obj.height + 2.*bufferSize;
            colFrame = obj.width + 2.*bufferSize;
            [rFrame, cFrame] = meshgrid(1:colFrame, 1:rowFrame);
            
            % numDots: # of coherently moving dots in the stimulus
            % numDotsNonCoh: # of noncoherently moving dots
            numDots = round(coherence.*density.*numel(cFrame));
            numDotsNonCoh = round((1-coherence).*density.*numel(cFrame));
            
            % Set the initial dot positions.
            % dotPositions is a matrix of positions of the coherently 
            % moving dots in [x,y] coordinates; each row in dotPositions 
            % stores the position of one dot
            if strcmp(obj.densityStyle, 'exact')
                z = zeros(numel(cFrame), 1);
                z(1:numDots) = 1;
                ord = rand(size(z));
                [~, ord] = sort(ord);
                z = z(ord);
                dotPos = [cFrame(z==1), rFrame(z==1)];
            else
                dotPos = rand(numDots, 2) * [rowFrame 0; 0 colFrame];
            end
            
            % s will store the output. After looping over each frame, we
            % will trim away the buffer from s to obtain the final result
            channel = zeros(rowFrame, colFrame, obj.channels);
            toInterp = -floor(1.5*dotRadius):floor(1.5*dotRadius);
            dSz = floor(1.5*dotRadius);
            
            for t=1:length
                % update the dot positions according to RDK type
                dotVel = [sin(dir/180*pi), cos(dir/180*pi)];
                dotVel = dotVel * speed;
                dotPos = dotPos + repmat(dotVel, size(dotPos, 1), 1);
                
                % wrap around for all dots that have gone past the image
                % borders
                w = find(dotPos(:,1) > rowFrame + 0.5);
                dotPos(w,1) = dotPos(w,1) - rowFrame;
                
                w = find(dotPos(:,1) < 0.5);
                dotPos(w,1) = dotPos(w,1) + rowFrame;
                
                w = find(dotPos(:,2) > colFrame + 0.5);
                dotPos(w,2) = dotPos(w,2) - colFrame;
                
                w = find(dotPos(:,2) < 0.5);
                dotPos(w,2) = dotPos(w,2) + colFrame;
                
                
                % add noncoherent dots and make a vector of dot positions 
                % for this frame only
                dotPosNonCoh = rand(numDotsNonCoh, 2) * ...
                    [rowFrame-1 0; 0 colFrame-1] + 0.5;
                
                % create a temporary matrix of positions for dots to be 
                % shown in this frame
                tmpDotPos = [dotPos; dotPosNonCoh];
                
                % prepare a matrix of zeros for this frame
                thisFrame = zeros(size(cFrame));
                if dotRadius > 0
                    % in each frame, don't show dots near the edges of the
                    % frame. That's why we have a buffer. The reason we
                    % don't show them is that we don't want to deal with
                    % edge handling
                    w1 = find(tmpDotPos(:,1) > (rowFrame - bufferSize + ...
                        + 1.5*dotRadius));
                    w2 = find(tmpDotPos(:,1) < (bufferSize - 1.5*dotRadius));
                    w3 = find(tmpDotPos(:,2) > (colFrame - bufferSize + ...
                        + 1.5*dotRadius));
                    w4 = find(tmpDotPos(:,2) < (bufferSize - 1.5*dotRadius));
                    w = [w1; w2; w3; w4];
                    tmpDotPos(w,:) = [];
                    
                    % add the dots to thisFrame
                    for p = 1:size(tmpDotPos, 1)
                        % find the center point of the current dot, in 
                        % thisFrame coordinates. This is where the dot will
                        % be placed.
                        cpX = round(tmpDotPos(p, 1));
                        cpY = round(tmpDotPos(p, 2));
                        
                        rToInterp = toInterp + (round(tmpDotPos(p,1)) - ...
                            tmpDotPos(p,1));
                        cToInterp = toInterp + (round(tmpDotPos(p,2)) - ...
                            tmpDotPos(p,2));
                        [rToInterp, cToInterp] = meshgrid(rToInterp, ...
                            cToInterp);
                        thisSmallDot = interp2(rLargeDot, cLargeDot, ...
                            largeDot, rToInterp, cToInterp, ...
                            obj.interpMethod);
                        
                        % now add this small dot to the frame.
                        thisFrame(cpX-dSz:cpX+dSz, cpY-dSz:cpY+dSz) = ...
                            thisFrame(cpX-dSz:cpX+dSz, cpY-dSz:cpY+dSz) ...
                            + thisSmallDot;
                        
                    end
                else
                    tmpDotPos(tmpDotPos(:,1) > rowFrame, :) = [];
                    tmpDotPos(tmpDotPos(:,1) < 1, :) = [];
                    tmpDotPos(tmpDotPos(:,2) > colFrame, :) = [];
                    tmpDotPos(tmpDotPos(:,2) < 1, :) = [];
                    tmpDotPos = round(tmpDotPos);
                    
                    w = sub2ind(size(thisFrame), tmpDotPos(:,1), ...
                        tmpDotPos(:,2));
                    
                    thisFrame(w) = 1;
                end
                % Add this frame to the final output
                channel(:,:,t) = thisFrame;
            end
            % Now trim away the buff
            channel = channel(bufferSize+1:end-bufferSize, ...
                bufferSize+1:end-bufferSize, :);
            channel(channel>1) = 1;
            
            % adjust for color
			if obj.channels == 1
				if mean(obj.colorVec) == 1
					% white is highest: stimulus unchanged
					dotCloud(:,:,1,:) = channel;
				else
					% reverse contrast
					dotCloud(:,:,1,:) = -channel + 1;
				end
			else
				sz = size(channel);
				dotCloud = zeros([sz(1:2) obj.channels sz(3)]);
				for c=1:obj.channels
					dotCloud(:,:,c,:) = channel * obj.colorVec(c);
				end
            end
            
            dotCloud = min(1, max(0, dotCloud));
            
            % use append method from base class to append frames
            obj.appendFrames(dotCloud);
        end
    end
    
    %% Protected Methods
    methods (Access = protected)
        function initDefaultParamsDerived(obj)
            obj.dotDirection = 0;
            obj.dotSpeed = 1;
            obj.dotDensity = 0.1;
            obj.dotCoherence = 0.5;
            obj.dotRadius = -1;
            obj.densityStyle = 'random';
            obj.sampleFactor = 10;
            obj.interpMethod = 'linear';

            obj.baseMsgId = 'VisualStimulus:DotStim';
            obj.name = 'DotStim';
			obj.stimType = eval(['obj.supportedStimTypes.' obj.name]);
			obj.colorChar = 'w';
			obj.colorVec = obj.char2rgb(obj.colorChar);
        end
    end
    
    %% Properties
    properties (Hidden, Access = protected)
        baseMsgId;          % string prepended to error messages
        name;               % string describing the stimulus type
		colorChar;          % single-character specifying stimulus color
		colorVec;           % 3-element vector specifying stimulus color
		stimType;           % integer from obj.supportedStimTypes
    end
    
    properties (Hidden, Access = private)
        dotDirection;       % drift direction
        dotSpeed;           % drift speed
        dotDensity;         % dot density
        dotCoherence;       % dot coherence
        dotRadius;          % dot size
        densityStyle;       % place dots randomly, although some dots might
                            % overlap
        sampleFactor;       % Dots are made by computing one large Gaussian
                            % dot, and interpolating from that dot to make
                            % smaller dots. This param specifies how much
                            % larger (a factor) the large dot compared to
                            % the smaller dots
        interpMethod;       % interpolation method: 'linear', 'cubic', 
                            % 'nearest', or 'v4'
    end
end
